physicalplan: minor fixes to pooling of FlowSpecs

yuzefovich opened the pull request:
**physicalplan: ensure that FlowSpecs are released**

We pool `FlowSpec` allocations in `GenerateFlowSpecs`, but previously we would only release them back to the pool on the main query path. This commit fixes that oversight.

**physicalplan: fix minor leak with pooling of FlowSpecs**

The only thing that we reuse when pooling `FlowSpec` objects is the capacity of `Processors` slice. Previously, we forgot to unset each element of that slice, so we could have hold onto to some processor specs which in turn could have hold onto some large objects (like `roachpb.Span`s in the TableReader) until a particular index of the slice is overwritten. This oversight is now fixed.

Found while looking at #140326.
Epic: None
Release note: None

yuzefovich committed:
physicalplan: ensure that FlowSpecs are released

We pool `FlowSpec` allocations in `GenerateFlowSpecs`, but previously we
would only release them back to the pool on the main query path. This
commit fixes that oversight.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141044)
<!-- Reviewable:end -->

yuzefovich committed:
physicalplan: fix minor leak with pooling of FlowSpecs

The only thing that we reuse when pooling `FlowSpec` objects is the
capacity of `Processors` slice. Previously, we forgot to unset each
element of that slice, so we could have hold onto to some processor
specs which in turn could have hold onto some large objects (like
`roachpb.Span`s in the TableReader) until a particular index of the
slice is overwritten. This oversight is now fixed.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141044#-:-OIqC5F8CYOMvGcOdLwy:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2)

___
*[`pkg/sql/physicalplan/physical_plan.go` line 985 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141044#-OIqBmtJ5gRfQvMhO2OV:-OIqBmtK5ey9_Eh7BpoS:b1zcnse) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e1061666377154cb0dcb6f67b6bcdec9278168/pkg/sql/physicalplan/physical_plan.go#L985)):*
> ```Go
> // The returned function should be called whenever the caller is done with all
> // FlowSpecs. The caller is free to ignore it if the specs will be released
> // separately.
> ```

Is the reason for returning the function instead of having an exported `ReleaseFlowSpecs` to force callers to consider cleanup?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141044#-:-OIqJ9gY5bw2uAyKvdTJ:b-oduykh)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @michae2)

___
*[`pkg/sql/physicalplan/physical_plan.go` line 985 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141044#-OIqBmtJ5gRfQvMhO2OV:-OIqIQDs5hFQyYQqrHcs:b-ujymzd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a9e1061666377154cb0dcb6f67b6bcdec9278168/pkg/sql/physicalplan/physical_plan.go#L985)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Is the reason for returning the function instead of having an exported `ReleaseFlowSpecs` to force callers to consider cleanup?
</blockquote></details>

Yeah, exactly - I want to make sure that every caller thinks about the cleanup, and pretty much in all cases the cleanup is deferred right away.


<!-- Sent from Reviewable.io -->
